### PR TITLE
Refactoring the refactor (II): Add artifact creation page

### PR DESCRIPTION
**Pull request by josenavas**
This depends on #1679, so review/merge that one first

This PR adds the page and code needed for adding an artifact to the system. At this point, this page is not integrated with the rest of the system in order to keep the PR small. It will be integrated in the next PR, which I'm almost done.

The tests, however, should pass.

The previous implementation was buggy and it did not allow to upload an artifact with more than a single lane or if it was missing the reverse reads. The user feedback has been also improved by performing the files checks on the fly rather when it clicks the create button.

### CODE REVIEW

**Comment by josenavas** [comment]
This is ready for review

File: qiita_pet/handlers/api_proxy/artifact.py [hunk 1]
```diff
@@ -73,8 +76,9 @@ def artifact_post_req(user_id, filepaths, artifact_type, name,
     ----------
     user_id : str
         User adding the atrifact
-    filepaths : dict of {str: [str, ...], ...}
-        List of files to attach to the artifact, keyed to the file type
+    filepaths : dict of {str: str}
```
Comment (antgonza) [inline qiita_pet/handlers/api_proxy/artifact.py]: Not sure the description and the example match.

File: qiita_pet/handlers/api_proxy/artifact.py [hunk 2]
```diff
@@ -92,23 +96,46 @@ def artifact_post_req(user_id, filepaths, artifact_type, name,
...
-                return {'status': 'error',
-                        'message': 'File does not exist: %s' % fp}
-            cleaned_filepaths.append((full_fp, ftype))
-
-    artifact = Artifact.create(cleaned_filepaths, artifact_type, name=name,
-                               prep_template=prep)
+    for ftype, file_list in viewitems(filepaths):
+        # JavaScript sends us this list as a comma-separated list
+        for fp in file_list.split(','):
+            # If the list was initially empty, the elemnt fp will the empty
```
Comment (antgonza) [inline qiita_pet/handlers/api_proxy/artifact.py]: element

File: qiita_pet/handlers/api_proxy/artifact.py [hunk 3]
```diff
@@ -92,23 +96,46 @@ def artifact_post_req(user_id, filepaths, artifact_type, name,
...
+            # string, which will make the check below for "exists" to check
```
Comment (antgonza) [inline qiita_pet/handlers/api_proxy/artifact.py]: can you rewrite? I don't think it makes a lot of sense.

File: qiita_pet/handlers/api_proxy/studies.py [hunk 4]
```diff
@@ -168,3 +171,72 @@ def study_prep_get_req(study_id, user_id):
...
+    study_id : int
+        The study id
+    prep_template_id : int
+        The prep template id
+    artifact_type : str
+        The artifact type
+
+    Returns
+    -------
+    dict of {str: object}
```
Comment (antgonza) [inline qiita_pet/handlers/api_proxy/studies.py]: {str: object} doesn't really match the below description.

File: qiita_pet/templates/study_ajax/add_artifact.html [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+        data: $("#create-artifact-form").serialize(),
+        success: function(data) {
+          if(data.status == 'error') {
+            bootstrapAlert(data.message, "danger");
+          }
+          else {
+            if(data.status == 'warning') {
+              bootstrapAlert(data.message, "warning");
+            }
+            // This reloads needs to be changed to:
```
Comment (antgonza) [inline qiita_pet/templates/study_ajax/add_artifact.html]: Just want to confirm when this section is gonna be updated, this or which PR?

**Comment by antgonza** [comment]
Some comments.

File: qiita_pet/handlers/api_proxy/artifact.py (hunk 1 above)
Comment (josenavas) [inline qiita_pet/handlers/api_proxy/artifact.py]: Good catch - done

File: qiita_pet/handlers/api_proxy/artifact.py (hunk 3 above)
Comment (josenavas) [inline qiita_pet/handlers/api_proxy/artifact.py]: Done

File: qiita_pet/handlers/api_proxy/studies.py (hunk 4 above)
Comment (josenavas) [inline qiita_pet/handlers/api_proxy/studies.py]: The returned dictionary is keyed by strings, but the values to those keys can be either strings, int, or lists. In order to englobe all these types in the type description, I used the common base class `object`. Do you have a better alternative?

File: qiita_pet/templates/study_ajax/add_artifact.html (hunk 5 above)
Comment (josenavas) [inline qiita_pet/templates/study_ajax/add_artifact.html]: Following PR

File: qiita_pet/handlers/api_proxy/artifact.py [hunk 6]
```diff
@@ -73,8 +76,9 @@ def artifact_post_req(user_id, filepaths, artifact_type, name,
     ----------
     user_id : str
         User adding the atrifact
-    filepaths : dict of {str: [str, ...], ...}
-        List of files to attach to the artifact, keyed to the file type
+    filepaths : str
```
Comment (ElDeveloper) [inline qiita_pet/handlers/api_proxy/artifact.py]: This description doesn't match "my inferred type", it seems like it really is a dictionary, not a string, am I missing something?

File: qiita_pet/handlers/api_proxy/artifact.py [hunk 7]
```diff
@@ -92,23 +96,48 @@ def artifact_post_req(user_id, filepaths, artifact_type, name,
...
+    # a explicit check, in case there is something odd with the JS
+    if not cleaned_filepaths:
+        return {'status': 'error',
+                'message': "Can't create artifact, no files provided."}
+
+    try:
+        artifact = Artifact.create(cleaned_filepaths, artifact_type, name=name,
+                                   prep_template=prep)
+    except Exception as e:
+        # We should hit this exception rarelly (that's why is an exception)
```
Comment (ElDeveloper) [inline qiita_pet/handlers/api_proxy/artifact.py]: rarelly -> rarely
why is -> why it is

File: qiita_pet/handlers/api_proxy/artifact.py [hunk 8]
```diff
@@ -92,23 +96,48 @@ def artifact_post_req(user_id, filepaths, artifact_type, name,
...
+        # since at this point we have done multiple checks. However, it can
+        # occur in weird cases, so better let the GUI know that this fail
```
Comment (ElDeveloper) [inline qiita_pet/handlers/api_proxy/artifact.py]: fail -> failed

File: qiita_pet/handlers/api_proxy/artifact.py [hunk 9]
```diff
@@ -92,23 +96,48 @@ def artifact_post_req(user_id, filepaths, artifact_type, name,
...
-    artifact = Artifact.create(cleaned_filepaths, artifact_type, name=name,
-                               prep_template=prep)
+    for ftype, file_list in viewitems(filepaths):
+        # JavaScript sends us this list as a comma-separated list
+        for fp in file_list.split(','):
+            # JavaScript will send this value as an empty string if the
+            # list of files was empty. In such case, the split will generate
+            # a single element containing the empty string. Check for that case
+            # here and, if fp is not the empty string, proceed to check if
+            # the file exists
```
Comment (ElDeveloper) [inline qiita_pet/handlers/api_proxy/artifact.py]: Thanks for the description here, very useful. Just to check, there's no possible way to get commas in these filenames right? I'm wondering if what we should be getting should be a JSON object, but that I guess gets more in the topic of the REST API etc.

File: qiita_pet/handlers/api_proxy/studies.py [hunk 10]
```diff
@@ -168,3 +171,72 @@ def study_prep_get_req(study_id, user_id):
...
+        for _, filename in uploaded:
+            if filename.startswith(prep_prefixes):
+                selected.append(filename)
+            else:
+                remaining.append(filename)
+    else:
+        num_prefixes = 0
+        remaining = [f for _, f in uploaded]
+
+    # At this point we can't make anything smart about selecting by default
```
Comment (ElDeveloper) [inline qiita_pet/handlers/api_proxy/studies.py]: make -> do

File: qiita_pet/handlers/api_proxy/studies.py [hunk 11]
```diff
@@ -168,3 +171,72 @@ def study_prep_get_req(study_id, user_id):
...
+    supp_file_types = supported_filepath_types(artifact_type)
+    selected = []
+    remaining = []
+
+    uploaded = get_files_from_uploads_folders(study_id)
+    pt = PrepTemplate(prep_template_id).to_dataframe()
+
+    if (any(ft.startswith('raw_') for ft, _ in supp_file_types) and
+            'run_prefix' in pt.columns):
+        prep_prefixes = tuple(set(pt['run_prefix']))
```
Comment (ElDeveloper) [inline qiita_pet/handlers/api_proxy/studies.py]: What about `tuple(pt['run_prefixes'].unique())`? Embracing pandas, not sure it's faster or anything, but the intent is clearer.

File: qiita_pet/templates/study_ajax/add_artifact.html [hunk 12]
```diff
@@ -0,0 +1,112 @@
+<script type="text/javascript">
+  // Global counter used to check if the form can be submitted or not
+  var error_count = 0;
```
Comment (ElDeveloper) [inline qiita_pet/templates/study_ajax/add_artifact.html]: Is there any way we can try an avoid adding global variables? I know it sometimes hard, but this is a slippery slope and it has the potential of getting out of control.

File: qiita_pet/templates/study_ajax/artifact_file_selector.html [hunk 13]
```diff
@@ -0,0 +1,186 @@
...
+   * Function that validates all the files lists
+   *
+   * This function triggers the sortupdate event in all the elements
+   * that are "checkeable"
+   */
+  function check_files_lists() {
+      // Reset the error counter
+      error_count = 0;
+      // initialize the error count to the number of file types
+      $(".checkeable").trigger("sortupdate");
```
Comment (ElDeveloper) [inline qiita_pet/templates/study_ajax/artifact_file_selector.html]: I just picked this up, but it seems a bit odd that the name of the class is `checkeable`, are you sure it's not `checkable`?

File: qiita_pet/templates/study_ajax/artifact_file_selector.html [hunk 14]
```diff
@@ -0,0 +1,186 @@
...
+   *
+   * This function checks that the contents of the sortable list is correct,
+   * taking into account if the lsit is required and the number of elements
+   * in the list.
+   *
+   */
+  function on_change_validation() {
+    var num_children = this.children.length;
+    var is_required = $(this).attr("data-file-required") === "True";
+    var req_count = +$(this).attr("data-file-count");
```
Comment (ElDeveloper) [inline qiita_pet/templates/study_ajax/artifact_file_selector.html]: What's the plus sign here for `+$`?

**Comment by ElDeveloper** [comment]
Just a few comments.

File: qiita_pet/handlers/api_proxy/artifact.py (hunk 6 above)
Comment (josenavas) [inline qiita_pet/handlers/api_proxy/artifact.py]: :beer: you're right

File: qiita_pet/handlers/api_proxy/artifact.py (hunk 9 above)
Comment (josenavas) [inline qiita_pet/handlers/api_proxy/artifact.py]: The serialization of that object happens in the HTML serialization of the form. I'm unsure why javascript (or HTML) is not serializing that list as a JSON object and self.argument is not unwrapping it like that, but that is why I'm receiving in the handler. Agree that we should totally revisit for a REST API

File: qiita_pet/handlers/api_proxy/artifact.py (hunk 7 above)
Comment (josenavas) [inline qiita_pet/handlers/api_proxy/artifact.py]: Done

File: qiita_pet/handlers/api_proxy/artifact.py (hunk 8 above)
Comment (josenavas) [inline qiita_pet/handlers/api_proxy/artifact.py]: Done

File: qiita_pet/handlers/api_proxy/studies.py (hunk 11 above)
Comment (josenavas) [inline qiita_pet/handlers/api_proxy/studies.py]: Pandas is ~2X slower so leaving as it is:
```
In [18]: %timeit tuple(pt['run_prefix'].unique())
10000 loops, best of 3: 19.3 µs per loop

In [19]: %timeit tuple(set(pt['run_prefix']))
100000 loops, best of 3: 9.15 µs per loop
```

File: qiita_pet/handlers/api_proxy/studies.py (hunk 10 above)
Comment (josenavas) [inline qiita_pet/handlers/api_proxy/studies.py]: Done

File: qiita_pet/templates/study_ajax/add_artifact.html (hunk 12 above)
Comment (josenavas) [inline qiita_pet/templates/study_ajax/add_artifact.html]: donezzo - some JS magic and done :P

File: qiita_pet/templates/study_ajax/artifact_file_selector.html (hunk 13 above)
Comment (josenavas) [inline qiita_pet/templates/study_ajax/artifact_file_selector.html]: Is a class that I made up, so yeah, probably the correct work is `checkable`.

File: qiita_pet/templates/study_ajax/artifact_file_selector.html (hunk 14 above)
Comment (josenavas) [inline qiita_pet/templates/study_ajax/artifact_file_selector.html]: explicit cast to int

File: qiita_pet/handlers/api_proxy/studies.py (hunk 4 above)
Comment (antgonza) [inline qiita_pet/handlers/api_proxy/studies.py]: What about: "see below"? No hard feelings so fine to leave it as is.

**Comment by antgonza** [comment]
:+1: @ElDeveloper can you merge?

File: qiita_pet/templates/study_ajax/artifact_file_selector.html (hunk 14 above)
Comment (ElDeveloper) [inline qiita_pet/templates/study_ajax/artifact_file_selector.html]: Thanks, I keep forgetting this.

File: qiita_pet/handlers/api_proxy/studies.py (hunk 11 above)
Comment (ElDeveloper) [inline qiita_pet/handlers/api_proxy/studies.py]: :+1: :hourglass_flowing_sand:

**Pull request merged**